# GLOBAL WEATHER DATA ANALYSIS

- This is an analysis project to understand the weather of countries in 2024

In [1]:
import pandas as pd # data handling
import matplotlib.pyplot as plt # visualization
import seaborn as sns # visualization

- pandas for data manipulation
- matplotlib and seaborn for visualization

## LOADING DATASET

In [2]:
df=pd.read_csv("weather.csv")
df

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,4:50 AM,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,5:21 AM,7:54 PM,12:58 PM,2:14 AM,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 9:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,5:40 AM,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,6:31 AM,9:11 PM,2:12 PM,3:31 AM,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 9:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,6:12 AM,5:55 PM,1:17 PM,12:38 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 2:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,6:14 AM,6:39 PM,8:54 AM,10:12 PM,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,5:34 AM,6:17 PM,7:30 AM,9:34 PM,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 9:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,5:46 AM,6:18 PM,8:00 AM,9:39 PM,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 8:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,6:15 AM,5:56 PM,9:13 AM,8:35 PM,Waxing Crescent,9


## inspecting data

In [3]:
df.head()

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.52,69.18,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.4,26.6,1,1,4:50 AM,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55
1,Albania,Tirana,41.33,19.82,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.1,2.0,1,1,5:21 AM,7:54 PM,12:58 PM,2:14 AM,Waxing Gibbous,55
2,Algeria,Algiers,36.76,3.05,Africa/Algiers,1715849100,2024-05-16 9:45,23.0,73.4,Sunny,...,10.4,18.4,1,1,5:40 AM,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.50,1.52,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.7,0.9,1,1,6:31 AM,9:11 PM,2:12 PM,3:31 AM,Waxing Gibbous,55
4,Angola,Luanda,-8.84,13.23,Africa/Luanda,1715849100,2024-05-16 9:45,26.0,78.8,Partly cloudy,...,183.4,262.3,5,10,6:12 AM,5:55 PM,1:17 PM,12:38 AM,Waxing Gibbous,55


In [4]:
df.tail()

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 2:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,6:14 AM,6:39 PM,8:54 AM,10:12 PM,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,5:34 AM,6:17 PM,7:30 AM,9:34 PM,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 9:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,5:46 AM,6:18 PM,8:00 AM,9:39 PM,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 8:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,6:15 AM,5:56 PM,9:13 AM,8:35 PM,Waxing Crescent,9
136632,Zimbabwe,Harare,-17.8178,31.0447,Africa/Harare,1776666600,2026-04-20 8:30,20.0,68.1,Sunny,...,7.85,8.05,1,1,6:06 AM,5:43 PM,9:07 AM,8:18 PM,Waxing Crescent,9


In [5]:
df.sample()

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
126290,Austria,Vienna,48.2,16.3667,Europe/Vienna,1772088300,2026-02-26 7:45,2.0,35.6,Sunny,...,26.05,33.05,2,3,6:42 AM,5:34 PM,11:16 AM,3:52 AM,Waxing Gibbous,66


## checking data types

In [6]:
df.dtypes

country                          object
location_name                    object
latitude                        float64
longitude                       float64
timezone                         object
last_updated_epoch                int64
last_updated                     object
temperature_celsius             float64
temperature_fahrenheit          float64
condition_text                   object
wind_mph                        float64
wind_kph                        float64
wind_degree                       int64
wind_direction                   object
pressure_mb                       int64
pressure_in                     float64
precip_mm                       float64
precip_in                       float64
humidity                          int64
cloud                             int64
feels_like_celsius              float64
feels_like_fahrenheit           float64
visibility_km                   float64
visibility_miles                  int64
uv_index                        float64


In [7]:
df.shape

(136633, 41)

In [8]:
df.columns

Index(['country', 'location_name', 'latitude', 'longitude', 'timezone',
       'last_updated_epoch', 'last_updated', 'temperature_celsius',
       'temperature_fahrenheit', 'condition_text', 'wind_mph', 'wind_kph',
       'wind_degree', 'wind_direction', 'pressure_mb', 'pressure_in',
       'precip_mm', 'precip_in', 'humidity', 'cloud', 'feels_like_celsius',
       'feels_like_fahrenheit', 'visibility_km', 'visibility_miles',
       'uv_index', 'gust_mph', 'gust_kph', 'air_quality_Carbon_Monoxide',
       'air_quality_Ozone', 'air_quality_Nitrogen_dioxide',
       'air_quality_Sulphur_dioxide', 'air_quality_PM2.5', 'air_quality_PM10',
       'air_quality_us-epa-index', 'air_quality_gb-defra-index', 'sunrise',
       'sunset', 'moonrise', 'moonset', 'moon_phase', 'moon_illumination'],
      dtype='object')

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136633 entries, 0 to 136632
Data columns (total 41 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   country                       136633 non-null  object 
 1   location_name                 136633 non-null  object 
 2   latitude                      136633 non-null  float64
 3   longitude                     136633 non-null  float64
 4   timezone                      136633 non-null  object 
 5   last_updated_epoch            136633 non-null  int64  
 6   last_updated                  136633 non-null  object 
 7   temperature_celsius           136633 non-null  float64
 8   temperature_fahrenheit        136633 non-null  float64
 9   condition_text                136633 non-null  object 
 10  wind_mph                      136633 non-null  float64
 11  wind_kph                      136633 non-null  float64
 12  wind_degree                   136633 non-nul

## informations

- number of columns = 41
- number of rows = 136633
- missing values = 0
- memory usage = 42.7+ MB
- sunrise,moonset and sunset times stored as object
- last_update stored as object

## Data types

- float = 21
- integer = 9
- objects = 11

## selecting columns

- selecting sinle columns

In [10]:
df['country']

0         Afghanistan
1             Albania
2             Algeria
3             Andorra
4              Angola
             ...     
136628      Venezuela
136629        Vietnam
136630          Yemen
136631         Zambia
136632       Zimbabwe
Name: country, Length: 136633, dtype: object

In [11]:
df[['sunrise','sunset']]

,sunrise,sunset
0,4:50 AM,6:50 PM
1,5:21 AM,7:54 PM
2,5:40 AM,7:50 PM
3,6:31 AM,9:11 PM
4,6:12 AM,5:55 PM
...,...,...
136628,6:14 AM,6:39 PM
136629,5:34 AM,6:17 PM
136630,5:46 AM,6:18 PM
136631,6:15 AM,5:56 PM


In [12]:
place=df[['country','location_name']]

In [13]:
place

,country,location_name
0,Afghanistan,Kabul
1,Albania,Tirana
2,Algeria,Algiers
3,Andorra,Andorra La Vella
4,Angola,Luanda
...,...,...
136628,Venezuela,Caracas
136629,Vietnam,Hanoi
136630,Yemen,Sanaa
136631,Zambia,Lusaka


## filter the data

In [14]:
df[df['temperature_celsius']==25]

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
31,Cameroon,Yaounde,3.8700,11.5200,Africa/Douala,1715849100,2024-05-16 9:45,25.0,76.9,Partly Cloudy,...,25.60,31.20,2,3,6:01 AM,6:19 PM,1:13 PM,12:59 AM,Waxing Gibbous,55
43,Cyprus,Nicosia,35.1700,33.3700,Asia/Famagusta,1715849100,2024-05-16 11:45,25.0,77.0,Partly cloudy,...,7.70,13.50,1,1,5:42 AM,7:45 PM,1:11 PM,2:08 AM,Waxing Gibbous,55
82,Israel,Jerusalem,31.7800,35.2300,Asia/Jerusalem,1715849100,2024-05-16 11:45,25.0,77.0,Partly cloudy,...,5.00,9.40,1,1,5:42 AM,7:30 PM,1:07 PM,1:55 AM,Waxing Gibbous,55
144,Rwanda,Kigali,-1.9500,30.0600,Africa/Kigali,1715849100,2024-05-16 10:45,25.0,77.0,Partly cloudy,...,4.20,6.30,1,1,5:55 AM,5:57 PM,1:02 PM,12:36 AM,Waxing Gibbous,55
172,Tanzania,Dodoma,-6.1800,35.7500,Africa/Dar_es_Salaam,1715849100,2024-05-16 11:45,25.0,77.0,Partly cloudy,...,1.90,3.40,1,1,6:38 AM,6:28 PM,1:43 PM,1:08 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136101,Equatorial Guinea,Malabo,3.7500,8.7833,Africa/Malabo,1776493800,2026-04-18 7:30,25.0,77.0,Patchy light rain with thunder,...,6.65,9.75,1,1,6:18 AM,6:30 PM,6:54 AM,7:33 PM,Waxing Crescent,0
136389,Saint Lucia,Castries,14.0000,-61.0000,America/St_Lucia,1776579300,2026-04-19 2:15,25.0,77.0,Partly cloudy,...,8.55,13.35,1,1,5:48 AM,6:18 PM,7:22 AM,8:48 PM,Waxing Crescent,4
136390,Saint Vincent and the Grenadines,Kingstown,13.1333,-61.2167,America/St_Vincent,1776579300,2026-04-19 2:15,25.0,77.0,Partly cloudy,...,6.60,10.00,1,1,5:50 AM,6:18 PM,7:25 AM,8:47 PM,Waxing Crescent,4
136483,Democratic Republic of Congo,Kinshasa,-4.3000,15.3000,Africa/Kinshasa,1776666600,2026-04-20 7:30,25.0,77.1,Patchy rain nearby,...,14.75,14.95,1,2,5:58 AM,5:58 PM,8:42 AM,8:55 PM,Waxing Crescent,9


In [15]:
df[df['moon_phase']=='New Moon']

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
4285,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1717682400,2024-06-06 18:30,18.0,64.3,Patchy rain nearby,...,3.90,8.20,1,1,4:40 AM,7:04 PM,4:02 AM,7:18 PM,New Moon,0
4286,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1717682400,2024-06-06 16:00,32.0,89.6,Sunny,...,4.50,5.10,1,1,5:08 AM,8:11 PM,4:31 AM,8:41 PM,New Moon,0
4287,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1717682400,2024-06-06 15:00,32.0,89.6,Sunny,...,16.40,34.80,2,2,5:29 AM,8:04 PM,4:58 AM,8:33 PM,New Moon,0
4288,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1717682400,2024-06-06 16:00,22.5,72.6,Patchy rain nearby,...,2.20,3.40,1,1,6:17 AM,9:29 PM,5:43 AM,10:03 PM,New Moon,0
4289,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1717682400,2024-06-06 15:00,26.0,78.8,Partly cloudy,...,7.30,25.20,1,1,6:17 AM,5:55 PM,6:04 AM,5:55 PM,New Moon,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136196,Samoa,Apia,-13.8333,-171.7333,Pacific/Apia,1776493800,2026-04-18 19:30,27.4,81.3,Partly cloudy,...,4.65,6.75,1,1,6:33 AM,6:19 PM,6:50 AM,6:40 PM,New Moon,0
136207,Solomon Islands,Honiara,-9.4333,159.9500,Pacific/Guadalcanal,1776493800,2026-04-18 17:30,31.2,88.2,Patchy rain nearby,...,2.45,2.95,1,1,6:23 AM,6:16 PM,6:42 AM,6:44 PM,New Moon,0
136224,Tonga,Nuku`Aloia,-21.1333,-175.2000,Pacific/Tongatapu,1776493800,2026-04-18 19:30,24.1,75.4,Clear,...,6.15,9.45,1,1,6:53 AM,6:27 PM,7:14 AM,6:43 PM,New Moon,0
136229,Tuvalu,Funafuti,-8.5167,179.2167,Pacific/Funafuti,1776493800,2026-04-18 18:30,30.3,86.5,Partly cloudy,...,1.75,2.25,1,1,6:06 AM,5:59 PM,6:21 AM,6:25 PM,New Moon,0


In [16]:
df[df['condition_text']=='Sunny']

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 9:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,5:40 AM,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
12,Bahrain,Manama,26.2400,50.5800,Asia/Bahrain,1715849100,2024-05-16 11:45,36.0,96.8,Sunny,...,39.00,114.30,2,4,4:50 AM,6:18 PM,12:10 PM,12:45 AM,Waxing Gibbous,55
15,Belarus,Minsk,53.9000,27.5700,Europe/Minsk,1715849100,2024-05-16 11:45,17.0,62.6,Sunny,...,1.60,1.70,1,1,5:05 AM,9:08 PM,12:59 PM,3:13 AM,Waxing Gibbous,55
22,Botswana,Gaborone,-24.6500,25.9100,Africa/Gaborone,1715849100,2024-05-16 10:45,22.0,71.6,Sunny,...,16.40,24.80,2,2,6:45 AM,5:40 PM,1:40 PM,12:27 AM,Waxing Gibbous,55
28,Madagascar,Ivory,-24.3700,46.4500,Indian/Antananarivo,1715849100,2024-05-16 11:45,27.4,81.4,Sunny,...,0.50,0.80,1,1,6:23 AM,5:18 PM,1:16 PM,12:02 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136622,United Arab Emirates,Abu Dhabi,24.4667,54.3667,Asia/Dubai,1776666600,2026-04-20 10:30,27.4,81.3,Sunny,...,65.85,170.85,4,9,5:57 AM,6:47 PM,7:55 AM,10:19 PM,Waxing Crescent,8
136623,United Kingdom,London,51.5171,-0.1062,Europe/London,1776666600,2026-04-20 7:30,3.2,37.8,Sunny,...,14.35,18.85,1,2,5:54 AM,8:05 PM,6:54 AM,No moonset,Waxing Crescent,9
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 9:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,5:46 AM,6:18 PM,8:00 AM,9:39 PM,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 8:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,6:15 AM,5:56 PM,9:13 AM,8:35 PM,Waxing Crescent,9


In [17]:
df[df['moon_phase']=='Waxing Gibbous']

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,4:50 AM,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,5:21 AM,7:54 PM,12:58 PM,2:14 AM,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 9:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,5:40 AM,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,6:31 AM,9:11 PM,2:12 PM,3:31 AM,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 9:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,6:12 AM,5:55 PM,1:17 PM,12:38 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
132922,Vanuatu,Port Vila,-17.7333,168.3167,Pacific/Efate,1775025900,2026-04-01 17:45,28.1,82.6,Patchy rain nearby,...,8.55,13.65,1,1,5:53 AM,5:49 PM,5:09 PM,4:49 AM,Waxing Gibbous,97
132924,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1775025900,2026-04-01 13:45,26.0,78.8,Cloudy,...,27.85,29.35,2,3,5:50 AM,6:11 PM,5:35 PM,5:06 AM,Waxing Gibbous,98
132925,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1775025900,2026-04-01 9:45,20.0,67.9,Sunny,...,29.75,147.95,2,3,5:59 AM,6:15 PM,5:49 PM,5:19 AM,Waxing Gibbous,98
132926,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1775025900,2026-04-01 8:45,19.3,66.8,Cloudy,...,4.45,4.65,1,1,6:12 AM,6:09 PM,5:43 PM,5:29 AM,Waxing Gibbous,99


In [18]:
df[df['condition_text']=='Clear']

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
6,Argentina,Buenos Aires,-34.5900,-58.6700,America/Argentina/Buenos_Aires,1715849100,2024-05-16 5:45,8.0,46.4,Clear,...,4.000,5.300,1,1,7:43 AM,5:59 PM,2:36 PM,1:04 AM,Waxing Gibbous,55
8,Australia,Canberra,-35.2800,149.2200,Australia/Sydney,1715849100,2024-05-16 18:45,9.0,48.2,Clear,...,3.700,5.400,1,1,6:52 AM,5:07 PM,1:31 PM,No moonset,Waxing Gibbous,55
20,Bolivia,Sucre,-19.0400,-65.2600,America/La_Paz,1715849100,2024-05-16 4:45,12.3,54.1,Clear,...,4.600,9.100,1,1,6:41 AM,5:53 PM,1:48 PM,12:51 AM,Waxing Gibbous,55
35,Chile,Santiago,-33.4500,-70.6700,America/Santiago,1715849100,2024-05-16 4:45,1.0,33.8,Clear,...,211.100,268.600,5,10,7:28 AM,5:50 PM,2:23 PM,12:56 AM,Waxing Gibbous,55
42,Cuba,Havana,23.1300,-82.3600,America/Havana,1715849100,2024-05-16 4:45,26.0,78.8,Clear,...,4.700,18.400,1,1,6:47 AM,8:04 PM,2:24 PM,2:44 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136522,Jamaica,Kingston,18.0000,-76.8000,America/Jamaica,1776666600,2026-04-20 1:30,26.1,79.0,Clear,...,17.950,19.350,2,2,5:47 AM,6:25 PM,8:18 AM,10:11 PM,Waxing Crescent,11
136543,Mali,Bamako,12.6500,-8.0000,Africa/Bamako,1776666600,2026-04-20 6:30,27.4,81.3,Clear,...,28.950,109.150,2,3,6:17 AM,6:45 PM,8:43 AM,10:10 PM,Waxing Crescent,9
136574,Paraguay,Garrapata,-19.7000,-58.7167,America/Asuncion,1776666600,2026-04-20 3:30,26.2,79.1,Clear,...,5.478,5.578,1,1,7:06 AM,6:36 PM,10:30 AM,9:21 PM,Waxing Crescent,10
136575,Peru,Lima,-12.0500,-77.0500,America/Lima,1776666600,2026-04-20 1:30,21.4,70.5,Clear,...,17.050,30.450,2,2,6:14 AM,6:00 PM,9:26 AM,9:03 PM,Waxing Crescent,11


In [19]:
df[df['moon_phase']=='Waning Gibbous']

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
1751,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1716561000,2024-05-24 19:00,18.5,65.3,Patchy rain nearby,...,13.00,25.30,1,2,4:45 AM,6:56 PM,8:07 PM,4:50 AM,Waning Gibbous,100
1752,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1716561000,2024-05-24 16:30,26.0,78.8,Sunny,...,1.80,2.50,1,1,5:14 AM,8:01 PM,9:30 PM,5:20 AM,Waning Gibbous,100
1753,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1716561000,2024-05-24 15:30,34.0,93.2,Sunny,...,9.70,21.10,1,1,5:34 AM,7:56 PM,9:21 PM,5:46 AM,Waning Gibbous,100
1754,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1716561000,2024-05-24 16:30,10.3,50.5,Light rain shower,...,1.00,1.20,1,1,6:24 AM,9:19 PM,10:51 PM,6:30 AM,Waning Gibbous,100
1755,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1716561000,2024-05-24 15:30,29.0,84.2,Sunny,...,44.00,92.10,3,5,6:14 AM,5:54 PM,6:43 PM,6:52 AM,Waning Gibbous,100
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
134482,Vanuatu,Port Vila,-17.7333,168.3167,Pacific/Efate,1775717100,2026-04-09 17:45,27.1,80.8,Partly cloudy,...,4.85,7.25,1,1,5:55 AM,5:42 PM,10:52 PM,11:45 AM,Waning Gibbous,66
134484,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1775717100,2026-04-09 13:45,32.4,90.3,Partly cloudy,...,40.15,42.65,2,4,5:43 AM,6:14 PM,No moonrise,10:34 AM,Waning Gibbous,64
134485,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1775717100,2026-04-09 9:45,18.9,66.0,Sunny,...,11.85,25.75,1,1,5:53 AM,6:17 PM,No moonrise,11:04 AM,Waning Gibbous,63
134486,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1775717100,2026-04-09 8:45,16.8,62.3,Patchy rain nearby,...,2.25,2.35,1,1,6:13 AM,6:03 PM,11:39 PM,12:18 PM,Waning Gibbous,62


## filtering with multiple conditions

In [20]:
df[(df['moon_phase']=='New Moon') & (df['temperature_celsius']==25)]

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
4319,China,Beijing,39.9300,116.3900,Asia/Shanghai,1717682400,2024-06-06 22:00,25.0,77.0,Clear,...,105.400,120.600,4,10,4:46 AM,7:41 PM,3:59 AM,7:50 PM,New Moon,0
4370,Kazakhstan,Astana,51.1800,71.4300,Asia/Almaty,1717682400,2024-06-06 19:00,25.0,77.0,Moderate or heavy rain with thunder,...,0.700,0.800,1,1,4:01 AM,8:25 PM,3:11 AM,8:59 PM,New Moon,0
4381,Liechtenstein,Vaduz,47.1300,9.5200,Europe/Vaduz,1717682400,2024-06-06 16:00,25.0,77.0,Sunny,...,1.000,1.200,1,1,5:28 AM,9:14 PM,4:48 AM,9:53 PM,New Moon,0
4404,Nepal,Kathmandu,27.7200,85.3200,Asia/Kathmandu,1717682400,2024-06-06 19:45,25.0,77.0,Partly cloudy,...,26.000,42.400,2,3,5:08 AM,6:58 PM,4:29 AM,7:04 PM,New Moon,0
4409,Nigeria,Abuja,9.1800,7.1800,Africa/Lagos,1717682400,2024-06-06 15:00,25.0,77.0,Partly cloudy,...,2.200,4.200,1,1,6:11 AM,6:49 PM,5:52 AM,7:00 PM,New Moon,0
4415,Palau,Adkip,7.3600,134.5100,Pacific/Palau,1717682400,2024-06-06 23:00,25.0,77.0,Light rain,...,0.500,0.600,1,1,5:45 AM,6:17 PM,5:06 AM,6:05 PM,New Moon,0
4421,Philippines,Manila,14.6000,120.9800,Asia/Manila,1717682400,2024-06-06 22:00,25.0,77.0,Moderate rain,...,51.400,71.200,3,6,5:26 AM,6:23 PM,4:47 AM,6:17 PM,New Moon,0
4442,Solomon Islands,Honiara,-9.4300,159.9500,Pacific/Guadalcanal,1717682400,2024-06-07 1:00,25.0,77.0,Partly cloudy,...,1.400,6.400,1,1,6:31 AM,6:07 PM,5:54 AM,5:43 PM,New Moon,0
15338,France,Paris,48.8700,2.3300,Europe/Paris,1722774600,2024-08-04 14:30,25.0,77.1,Sunny,...,4.500,8.800,1,1,6:28 AM,9:24 PM,5:57 AM,9:53 PM,New Moon,0
20982,Dominican Republic,Santo Domingo,18.4700,-69.9000,America/Santo_Domingo,1725279300,2024-09-02 8:15,25.0,77.0,Partly cloudy,...,2.100,5.700,1,1,6:25 AM,6:52 PM,5:59 AM,6:51 PM,New Moon,1


## selecting rows - iloc method



In [21]:
df.iloc[9] # single row
df.iloc[4:11] # portion

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
4,Angola,Luanda,-8.84,13.23,Africa/Luanda,1715849100,2024-05-16 9:45,26.0,78.8,Partly cloudy,...,183.4,262.3,5,10,6:12 AM,5:55 PM,1:17 PM,12:38 AM,Waxing Gibbous,55
5,Antigua and Barbuda,Saint John's,17.12,-61.85,America/Antigua,1715849100,2024-05-16 4:45,26.0,78.8,Partly cloudy,...,1.2,4.5,1,1,5:36 AM,6:32 PM,1:05 PM,1:14 AM,Waxing Gibbous,55
6,Argentina,Buenos Aires,-34.59,-58.67,America/Argentina/Buenos_Aires,1715849100,2024-05-16 5:45,8.0,46.4,Clear,...,4.0,5.3,1,1,7:43 AM,5:59 PM,2:36 PM,1:04 AM,Waxing Gibbous,55
7,Armenia,Yerevan,40.18,44.51,Asia/Yerevan,1715849100,2024-05-16 12:45,19.0,66.2,Partly cloudy,...,0.8,0.9,1,1,5:45 AM,8:12 PM,1:17 PM,2:31 AM,Waxing Gibbous,55
8,Australia,Canberra,-35.28,149.22,Australia/Sydney,1715849100,2024-05-16 18:45,9.0,48.2,Clear,...,3.7,5.4,1,1,6:52 AM,5:07 PM,1:31 PM,No moonset,Waxing Gibbous,55
9,Austria,Vienna,48.20,16.37,Europe/Vienna,1715849100,2024-05-16 10:45,16.0,60.8,Partly cloudy,...,3.7,4.4,1,1,5:14 AM,8:29 PM,1:00 PM,2:42 AM,Waxing Gibbous,55
10,Azerbaijan,Baku,40.40,49.88,Asia/Baku,1715849100,2024-05-16 12:45,17.0,62.6,Partly cloudy,...,1.9,2.2,1,1,5:23 AM,7:51 PM,12:54 PM,2:10 AM,Waxing Gibbous,55


In [22]:
df.iloc[9]

country                                  Austria
location_name                             Vienna
latitude                                    48.2
longitude                                  16.37
timezone                           Europe/Vienna
last_updated_epoch                    1715849100
last_updated                    2024-05-16 10:45
temperature_celsius                         16.0
temperature_fahrenheit                      60.8
condition_text                     Partly cloudy
wind_mph                                    12.5
wind_kph                                    20.2
wind_degree                                  110
wind_direction                               ESE
pressure_mb                                 1013
pressure_in                                29.91
precip_mm                                    0.0
precip_in                                    0.0
humidity                                      63
cloud                                         75
feels_like_celsius  

## loc method

In [23]:
df.loc[100,'sunrise']

'5:49 AM'

## rename column

In [24]:
df.rename(columns={"country":"place"})

,place,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temperature_celsius,temperature_fahrenheit,condition_text,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,4:50 AM,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,5:21 AM,7:54 PM,12:58 PM,2:14 AM,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 9:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,5:40 AM,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,6:31 AM,9:11 PM,2:12 PM,3:31 AM,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 9:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,6:12 AM,5:55 PM,1:17 PM,12:38 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 2:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,6:14 AM,6:39 PM,8:54 AM,10:12 PM,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,5:34 AM,6:17 PM,7:30 AM,9:34 PM,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 9:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,5:46 AM,6:18 PM,8:00 AM,9:39 PM,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 8:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,6:15 AM,5:56 PM,9:13 AM,8:35 PM,Waxing Crescent,9


In [25]:
df.rename(columns={"condition_text":"weather","temperature_celsius":"temp_celsius"},inplace=True)

In [26]:
df

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temp_celsius,temperature_fahrenheit,weather,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,4:50 AM,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,5:21 AM,7:54 PM,12:58 PM,2:14 AM,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 9:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,5:40 AM,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,6:31 AM,9:11 PM,2:12 PM,3:31 AM,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 9:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,6:12 AM,5:55 PM,1:17 PM,12:38 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 2:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,6:14 AM,6:39 PM,8:54 AM,10:12 PM,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,5:34 AM,6:17 PM,7:30 AM,9:34 PM,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 9:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,5:46 AM,6:18 PM,8:00 AM,9:39 PM,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 8:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,6:15 AM,5:56 PM,9:13 AM,8:35 PM,Waxing Crescent,9


In [27]:
df['weather'].value_counts()

weather
Partly cloudy                                  39376
Sunny                                          39229
Patchy rain nearby                             11376
Partly Cloudy                                   8669
Overcast                                        7410
Clear                                           7200
Mist                                            5716
Light rain                                      4287
Light rain shower                               3133
Fog                                             1709
Cloudy                                          1633
Moderate or heavy rain with thunder             1090
Moderate rain                                   1042
Patchy light rain with thunder                   843
Light drizzle                                    678
Light snow                                       523
Patchy light drizzle                             455
Moderate or heavy rain shower                    287
Thundery outbreaks in nearby          

In [28]:
df['moon_phase'].value_counts()

moon_phase
Waning Crescent    29886
Waxing Crescent    29481
Waxing Gibbous     29382
Waning Gibbous     28477
Full Moon           4933
Last Quarter        4920
New Moon            4876
First Quarter       4678
Name: count, dtype: int64

## Data cleaning

### Clean Numericals columns with characters

In [29]:
df['sunrise']=pd.to_numeric(
    df['sunrise']
    .astype(str)
    .str.replace(r'[^0-9,-]','',regex=True)
    .replace('',None),
    errors='coerce'
)

In [30]:
df

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temp_celsius,temperature_fahrenheit,weather,...,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunrise,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15,26.6,79.8,Partly Cloudy,...,8.40,26.60,1,1,450,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45,19.0,66.2,Partly cloudy,...,1.10,2.00,1,1,521,7:54 PM,12:58 PM,2:14 AM,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 9:45,23.0,73.4,Sunny,...,10.40,18.40,1,1,540,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45,6.3,43.3,Light drizzle,...,0.70,0.90,1,1,631,9:11 PM,2:12 PM,3:31 AM,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 9:45,26.0,78.8,Partly cloudy,...,183.40,262.30,5,10,612,5:55 PM,1:17 PM,12:38 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 2:30,17.5,63.6,Partly Cloudy,...,7.85,9.35,1,1,614,6:39 PM,8:54 AM,10:12 PM,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30,28.3,82.9,Cloudy,...,36.25,39.45,2,4,534,6:17 PM,7:30 AM,9:34 PM,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 9:30,21.0,69.7,Sunny,...,12.85,51.65,1,2,546,6:18 PM,8:00 AM,9:39 PM,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 8:30,21.5,70.8,Sunny,...,5.55,5.85,1,1,615,5:56 PM,9:13 AM,8:35 PM,Waxing Crescent,9


In [33]:
df["last_updated"] = pd.to_datetime(df["last_updated"],
                                    errors="coerce")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136633 entries, 0 to 136632
Data columns (total 41 columns):
 #   Column                        Non-Null Count   Dtype         
---  ------                        --------------   -----         
 0   country                       136633 non-null  object        
 1   location_name                 136633 non-null  object        
 2   latitude                      136633 non-null  float64       
 3   longitude                     136633 non-null  float64       
 4   timezone                      136633 non-null  object        
 5   last_updated_epoch            136633 non-null  int64         
 6   last_updated                  136633 non-null  datetime64[ns]
 7   temp_celsius                  136633 non-null  float64       
 8   temperature_fahrenheit        136633 non-null  float64       
 9   weather                       136633 non-null  object        
 10  wind_mph                      136633 non-null  float64       
 11  wind_kph     

In [40]:
df.dtypes

country                                 object
location_name                           object
latitude                               float64
longitude                              float64
timezone                                object
last_updated_epoch                       int64
last_updated                    datetime64[ns]
temp_celsius                           float64
temperature_fahrenheit                 float64
weather                                 object
wind_mph                               float64
wind_kph                               float64
wind_degree                              int64
wind_direction                          object
pressure_mb                              int64
pressure_in                            float64
precip_mm                              float64
precip_in                              float64
humidity                                 int64
cloud                                    int64
feels_like_celsius                     float64
feels_like_fa

## deleting unwanted columns

In [41]:
df.drop(columns=['sunrise'],inplace=True)

In [42]:
df

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temp_celsius,temperature_fahrenheit,weather,...,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index,sunset,moonrise,moonset,moon_phase,moon_illumination
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15:00,26.6,79.8,Partly Cloudy,...,0.20,8.40,26.60,1,1,6:50 PM,12:12 PM,1:11 AM,Waxing Gibbous,55
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45:00,19.0,66.2,Partly cloudy,...,0.10,1.10,2.00,1,1,7:54 PM,12:58 PM,2:14 AM,Waxing Gibbous,55
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45:00,23.0,73.4,Sunny,...,13.40,10.40,18.40,1,1,7:50 PM,1:15 PM,2:14 AM,Waxing Gibbous,55
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45:00,6.3,43.3,Light drizzle,...,0.20,0.70,0.90,1,1,9:11 PM,2:12 PM,3:31 AM,Waxing Gibbous,55
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45:00,26.0,78.8,Partly cloudy,...,31.50,183.40,262.30,5,10,5:55 PM,1:17 PM,12:38 AM,Waxing Gibbous,55
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30:00,17.5,63.6,Partly Cloudy,...,1.65,7.85,9.35,1,1,6:39 PM,8:54 AM,10:12 PM,Waxing Crescent,10
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30:00,28.3,82.9,Cloudy,...,21.05,36.25,39.45,2,4,6:17 PM,7:30 AM,9:34 PM,Waxing Crescent,7
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30:00,21.0,69.7,Sunny,...,2.05,12.85,51.65,1,2,6:18 PM,8:00 AM,9:39 PM,Waxing Crescent,8
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30:00,21.5,70.8,Sunny,...,2.05,5.55,5.85,1,1,5:56 PM,9:13 AM,8:35 PM,Waxing Crescent,9


In [43]:
df.drop(columns=['sunset','moonrise','moonset','moon_phase','moon_illumination'],inplace=True)

In [44]:
df

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temp_celsius,temperature_fahrenheit,weather,...,gust_mph,gust_kph,air_quality_Carbon_Monoxide,air_quality_Ozone,air_quality_Nitrogen_dioxide,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15:00,26.6,79.8,Partly Cloudy,...,9.5,15.3,277.00,103.0,1.10,0.20,8.40,26.60,1,1
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45:00,19.0,66.2,Partly cloudy,...,11.4,18.4,193.60,97.3,0.90,0.10,1.10,2.00,1,1
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45:00,23.0,73.4,Sunny,...,13.9,22.3,540.70,12.2,65.10,13.40,10.40,18.40,1,1
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45:00,6.3,43.3,Light drizzle,...,8.5,13.7,170.20,64.4,1.60,0.20,0.70,0.90,1,1
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45:00,26.0,78.8,Partly cloudy,...,12.5,20.2,2964.00,19.0,72.70,31.50,183.40,262.30,5,10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30:00,17.5,63.6,Partly Cloudy,...,6.7,10.8,577.85,35.0,8.15,1.65,7.85,9.35,1,1
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30:00,28.3,82.9,Cloudy,...,12.7,20.5,351.85,35.0,34.45,21.05,36.25,39.45,2,4
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30:00,21.0,69.7,Sunny,...,5.7,9.1,149.85,58.0,5.85,2.05,12.85,51.65,1,2
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30:00,21.5,70.8,Sunny,...,11.1,17.8,160.85,40.0,2.85,2.05,5.55,5.85,1,1


In [45]:
df.drop(columns=['temperature_fahrenheit','wind_mph','precip_in','visibility_miles','gust_mph','pressure_in','feels_like_fahrenheit'],inplace=True)

In [46]:
df

,country,location_name,latitude,longitude,timezone,last_updated_epoch,last_updated,temp_celsius,weather,wind_kph,...,uv_index,gust_kph,air_quality_Carbon_Monoxide,air_quality_Ozone,air_quality_Nitrogen_dioxide,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,1715849100,2024-05-16 13:15:00,26.6,Partly Cloudy,13.3,...,7.0,15.3,277.00,103.0,1.10,0.20,8.40,26.60,1,1
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,1715849100,2024-05-16 10:45:00,19.0,Partly cloudy,11.2,...,5.0,18.4,193.60,97.3,0.90,0.10,1.10,2.00,1,1
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,1715849100,2024-05-16 09:45:00,23.0,Sunny,15.1,...,5.0,22.3,540.70,12.2,65.10,13.40,10.40,18.40,1,1
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,1715849100,2024-05-16 10:45:00,6.3,Light drizzle,11.9,...,2.0,13.7,170.20,64.4,1.60,0.20,0.70,0.90,1,1
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,1715849100,2024-05-16 09:45:00,26.0,Partly cloudy,13.0,...,8.0,20.2,2964.00,19.0,72.70,31.50,183.40,262.30,5,10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,1776666600,2026-04-20 02:30:00,17.5,Partly Cloudy,3.6,...,0.0,10.8,577.85,35.0,8.15,1.65,7.85,9.35,1,1
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,1776666600,2026-04-20 13:30:00,28.3,Cloudy,17.6,...,5.5,20.5,351.85,35.0,34.45,21.05,36.25,39.45,2,4
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,1776666600,2026-04-20 09:30:00,21.0,Sunny,7.9,...,6.5,9.1,149.85,58.0,5.85,2.05,12.85,51.65,1,2
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,1776666600,2026-04-20 08:30:00,21.5,Sunny,13.0,...,1.6,17.8,160.85,40.0,2.85,2.05,5.55,5.85,1,1


In [49]:
df.shape

(136633, 28)

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136633 entries, 0 to 136632
Data columns (total 28 columns):
 #   Column                        Non-Null Count   Dtype         
---  ------                        --------------   -----         
 0   country                       136633 non-null  object        
 1   location_name                 136633 non-null  object        
 2   latitude                      136633 non-null  float64       
 3   longitude                     136633 non-null  float64       
 4   timezone                      136633 non-null  object        
 5   last_updated_epoch            136633 non-null  int64         
 6   last_updated                  136633 non-null  datetime64[ns]
 7   temp_celsius                  136633 non-null  float64       
 8   weather                       136633 non-null  object        
 9   wind_kph                      136633 non-null  float64       
 10  wind_degree                   136633 non-null  int64         
 11  wind_directio

## Handling missing values

In [52]:
df.isnull().sum()

country                         0
location_name                   0
latitude                        0
longitude                       0
timezone                        0
last_updated_epoch              0
last_updated                    0
temp_celsius                    0
weather                         0
wind_kph                        0
wind_degree                     0
wind_direction                  0
pressure_mb                     0
precip_mm                       0
humidity                        0
cloud                           0
feels_like_celsius              0
visibility_km                   0
uv_index                        0
gust_kph                        0
air_quality_Carbon_Monoxide     0
air_quality_Ozone               0
air_quality_Nitrogen_dioxide    0
air_quality_Sulphur_dioxide     0
air_quality_PM2.5               0
air_quality_PM10                0
air_quality_us-epa-index        0
air_quality_gb-defra-index      0
dtype: int64

- no missing values in this dataset
- filling with median(numerical) and with mode(categorical)

In [53]:
df['temp_celsius'].median()

23.9

In [54]:
df['temp_celsius'].mode()

0    25.1
Name: temp_celsius, dtype: float64

In [55]:
df.drop(columns=['last_updated_epoch'],inplace=True)

In [56]:
df

,country,location_name,latitude,longitude,timezone,last_updated,temp_celsius,weather,wind_kph,wind_degree,...,uv_index,gust_kph,air_quality_Carbon_Monoxide,air_quality_Ozone,air_quality_Nitrogen_dioxide,air_quality_Sulphur_dioxide,air_quality_PM2.5,air_quality_PM10,air_quality_us-epa-index,air_quality_gb-defra-index
0,Afghanistan,Kabul,34.5200,69.1800,Asia/Kabul,2024-05-16 13:15:00,26.6,Partly Cloudy,13.3,338,...,7.0,15.3,277.00,103.0,1.10,0.20,8.40,26.60,1,1
1,Albania,Tirana,41.3300,19.8200,Europe/Tirane,2024-05-16 10:45:00,19.0,Partly cloudy,11.2,320,...,5.0,18.4,193.60,97.3,0.90,0.10,1.10,2.00,1,1
2,Algeria,Algiers,36.7600,3.0500,Africa/Algiers,2024-05-16 09:45:00,23.0,Sunny,15.1,280,...,5.0,22.3,540.70,12.2,65.10,13.40,10.40,18.40,1,1
3,Andorra,Andorra La Vella,42.5000,1.5200,Europe/Andorra,2024-05-16 10:45:00,6.3,Light drizzle,11.9,215,...,2.0,13.7,170.20,64.4,1.60,0.20,0.70,0.90,1,1
4,Angola,Luanda,-8.8400,13.2300,Africa/Luanda,2024-05-16 09:45:00,26.0,Partly cloudy,13.0,150,...,8.0,20.2,2964.00,19.0,72.70,31.50,183.40,262.30,5,10
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
136628,Venezuela,Caracas,10.5000,-66.9167,America/Caracas,2026-04-20 02:30:00,17.5,Partly Cloudy,3.6,122,...,0.0,10.8,577.85,35.0,8.15,1.65,7.85,9.35,1,1
136629,Vietnam,Hanoi,21.0333,105.8500,Asia/Bangkok,2026-04-20 13:30:00,28.3,Cloudy,17.6,151,...,5.5,20.5,351.85,35.0,34.45,21.05,36.25,39.45,2,4
136630,Yemen,Sanaa,15.3547,44.2067,Asia/Aden,2026-04-20 09:30:00,21.0,Sunny,7.9,135,...,6.5,9.1,149.85,58.0,5.85,2.05,12.85,51.65,1,2
136631,Zambia,Lusaka,-15.4167,28.2833,Africa/Lusaka,2026-04-20 08:30:00,21.5,Sunny,13.0,58,...,1.6,17.8,160.85,40.0,2.85,2.05,5.55,5.85,1,1


- finding mean,median and mode of columns

In [58]:
print(df.select_dtypes(include='number').mean())

latitude                          19.211173
longitude                         21.956700
temp_celsius                      21.268191
wind_kph                          12.897045
wind_degree                      168.797348
pressure_mb                     1014.064304
precip_mm                          0.134168
humidity                          66.661553
cloud                             39.976228
feels_like_celsius                22.120190
visibility_km                      9.507450
uv_index                           3.313816
gust_kph                          18.242712
air_quality_Carbon_Monoxide      459.169559
air_quality_Ozone                 57.997988
air_quality_Nitrogen_dioxide      14.974117
air_quality_Sulphur_dioxide       10.315061
air_quality_PM2.5                 24.149276
air_quality_PM10                  48.293201
air_quality_us-epa-index           1.697430
air_quality_gb-defra-index         2.601275
dtype: float64


In [59]:
print(df.select_dtypes(include='number').median())

latitude                          17.2500
longitude                         23.2361
temp_celsius                      23.9000
wind_kph                          10.8000
wind_degree                      161.0000
pressure_mb                     1014.0000
precip_mm                          0.0000
humidity                          72.0000
cloud                             30.0000
feels_like_celsius                25.1000
visibility_km                     10.0000
uv_index                           1.9000
gust_kph                          15.3000
air_quality_Carbon_Monoxide      293.7000
air_quality_Ozone                 55.0000
air_quality_Nitrogen_dioxide       5.6500
air_quality_Sulphur_dioxide        2.4050
air_quality_PM2.5                 14.1500
air_quality_PM10                  19.9800
air_quality_us-epa-index           1.0000
air_quality_gb-defra-index         2.0000
dtype: float64


In [60]:
print(df.select_dtypes(include='object').mode())

     country location_name      timezone        weather wind_direction
0   Bulgaria         Accra  Asia/Bangkok  Partly cloudy              E
1        NaN         Amman           NaN            NaN            NaN
2        NaN        Asmara           NaN            NaN            NaN
3        NaN       Baghdad           NaN            NaN            NaN
4        NaN          Bern           NaN            NaN            NaN
5        NaN     Bujumbura           NaN            NaN            NaN
6        NaN         Dakar           NaN            NaN            NaN
7        NaN         Kabul           NaN            NaN            NaN
8        NaN          Kyiv           NaN            NaN            NaN
9        NaN        Malabo           NaN            NaN            NaN
10       NaN     N'djamena           NaN            NaN            NaN
11       NaN           Nan           NaN            NaN            NaN
12       NaN         Sanaa           NaN            NaN            NaN
13    